### PR TITLE
upgrade-2.x.sh: Add script to upgrade 2.x host OS

### PR DESCRIPTION
**Pull request by willnewton**
Add a script that can upgrade a 2.x resinOS version to another
2.x version.

### CODE REVIEW

File: upgrade-2.x.sh [hunk 1]
```diff
@@ -0,0 +1,201 @@
+#!/bin/bash
```
Comment (michal-mazurek) [inline upgrade-2.x.sh]: Bash scripts should have a .bash extension

File: upgrade-2.x.sh [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    fi
+}
+
+# Test if a version is greater than another
+function version_gt() {
+    test "$(echo "$@" | tr " " "\n" | sort -V | head -n 1)" != "$1"
+}
+
+# Log timer
+STARTTIME=$(date +%s)
```
Comment (michal-mazurek) [inline upgrade-2.x.sh]: non-exported functions should be lowercase, to indicate they are non-exported and make the code more readable.

File: upgrade-2.x.sh [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+# Check board support
+case $SLUG in
+    beaglebone*)
+	BINARY_TYPE=arm
```
Comment (michal-mazurek) [inline upgrade-2.x.sh]: Mixed tabs and spaces for indentation

File: upgrade-2.x.sh [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    2.*)
+	log "Host OS version \"$VERSION\" OK."
+	;;
+    *)
+	log ERROR "Host OS version \"$VERSION\" not supported."
+	;;
+esac
+
+# Check if we need to install some more extra tools
+if ! version_gt $VERSION $PREFERRED_HOSTOS_VERSION && ! [ "$VERSION" == $PREFERRED_HOSTOS_VERSION ]; then
```
Comment (michal-mazurek) [inline upgrade-2.x.sh]: Perhaps break long line

File: upgrade-2.x.sh [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+# Extract boot partition, exclude boot_whitelist files
+log "Extracting new boot partition..."
+echo resin-boot/cmdline.txt >/tmp/boot-exclude
+echo resin-boot/config.txt >>/tmp/boot-exclude
+echo resin-boot/splash/resin-logo.png >>/tmp/boot-exclude
+echo resin-boot/uEnv.txt >>/tmp/boot-exclude
+echo resin-boot/EFI/BOOT/grub.cfg >>/tmp/boot-exclude
+# 2.x adds a default config.json, we should avoid clobbering the existing one
+echo resin-boot/config.json >>/tmp/boot-exclude
```
Comment (michal-mazurek) [inline upgrade-2.x.sh]: Perhaps more elegant to write like this:
```
( echo
echo
echo) > /tmp/boot-exclude
```

File: upgrade-2.x.sh [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+	old_root=${root_part#/dev/}
+	new_root=${update_part#/dev/}
+	sed -i -e "s/$old_root/$new_root/" /mnt/boot/cmdline.txt
+	;;
+esac
+
+# Reboot into new OS
+sync
+log "Rebooting into new OS after 5 seconds ..."
+sleep 5
```
Comment (michal-mazurek) [inline upgrade-2.x.sh]: Why the delay? Perhaps a comment to explain.

Comment (willnewton) [inline upgrade-2.x.sh]: Mainly to give interested people the chance to Ctrl-C the script if they wish to. ;-) It's probably served it's purpose though and can be removed.

**Comment by willnewton** [comment]
@michal-mazurek I have rebased with changes based on your suggestions (apart from the file extension issue which there doesn't seem to be consensus on at the moment), thanks!

File: upgrade-2.x.sh [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        ;;
+    raspberry*)
+        binary_type=arm
+        ;;
+    *)
+        log ERROR "Unsupported board type $SLUG."
+esac
+
+if [ -n "$target_version" ]; then
+    target_version=$(echo "$target_version" | tr + _)
```
Comment (agherzan) [inline upgrade-2.x.sh]: A small explanation why this translation is needed would help (docker image name restrictions).

File: upgrade-2.x.sh [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        *)
+            log ERROR "Target OS version \"$target_version\" not supported."
+            ;;
+    esac
+else
+    log ERROR "No target OS version specified."
+fi
+
+# Check host OS version
+case $VERSION in
```
Comment (agherzan) [inline upgrade-2.x.sh]: I was little confused about who sets this variable. Maybe a comment that it comes from os-release would help. Or actually when sourcing we could say what variables we are relying on from that file. Just to be clear.

File: upgrade-2.x.sh [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        update_part=${root_dev}p2
+        update_part_no=2
+        update_label=resin-rootA
+        ;;
+    *)
+        log ERROR "Unknown root partition \"$root_part\"."
+esac
+
+# Stop docker containers
+log "Stopping all containers..."
```
Comment (agherzan) [inline upgrade-2.x.sh]: Can we avoid this now? Is supevisor still removing everything? @pcarranzav can we make it ignore resinos images?

File: upgrade-2.x.sh [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+EOF
+tar -x -X /tmp/root-exclude -C /tmp/updateroot -f $fsarchive
+
+# Extract quirks
+tar -x -C /tmp -f $fsarchive quirks
+cp -a /tmp/quirks/* /tmp/updateroot/
+rm -rf /tmp/quirks
+
+# XXX hack required until 2.0.7 when a quirk is added for this
+(cd /tmp/updateroot/etc; rm mtab; ln -s ../proc/self/mounts mtab)
```
Comment (agherzan) [inline upgrade-2.x.sh]: Should we only support updates to 2.0.7+? I don't see the benefit in updating to lower versions and keep this hack.

File: upgrade-2.x.sh (hunk 9 above)
Comment (willnewton) [inline upgrade-2.x.sh]: If we don't do that then something (I guess the supervisor) cleans up containers which I think can make the export of the container fail below.

Comment (agherzan) [inline upgrade-2.x.sh]: That is correct. I was wondering if we can make supervisor don't do that.

File: upgrade-2.x.sh (hunk 1 above)
Comment (petrosagg) [inline upgrade-2.x.sh]: I've never seen a `.bash` file. Out of all `.sh` files in my system's `/usr`, 40% are bash scripts

Comment (michal-mazurek) [inline upgrade-2.x.sh]: I posted a rationale for this once. Imagine you have the file open with a text editor. You then cannot execute it by running ./a.sh. So you run sh a.sh. Which then, perhaps only after executing half of the script, crashes with a syntax error.

**Pull request merged**